### DISCUSSION TITLE
Mutable data in the context?

### DISCUSSION THREAD

**Discussion by artsiomkaltovich** [async-graphql/async-graphql#1553] [Q&A]
Hello. Sorry if the query is dump, I am pretty new in rust and this library.

I am creating a simple graphql server and want to have a data provider which I can easily replace by in memory mock. So I have created this mock as an simple HashMap wrapper and it works perfectly. But I can't figure out how to make data mutable.

```rust
pub struct InMemoryDataProvider {
    models: HashMap<ModelId, Model>
}

impl DataProviderTrait for InMemoryDataProvider {
    async fn add_model(&mut self, model: Model) -> Result<ModelId, String> {
        let id = model.id;
        self.models.insert(id, model);
        Ok(id)
    }
}

pub enum DataProvider {
    InMemory(InMemoryDataProvider),
}

impl DataProvider {
    pub fn new() -> Self {
        Self::InMemory(InMemoryDataProvider::new())
    }

    pub async fn add_model(&mut self, model: Model) -> Result<ModelId, String> {
        info!("adding model {}", model.id);
        match self {
            Self::InMemory(data) => data.add_model(model).await,
        }
    }
}

#[Object]
impl Mutation {
    async fn add_model(&self, ctx: &Context<'_>, id: ModelId, name: String) -> Result<ModelId, String> {
        match ctx.data::<DataProvider>() {
            Ok(data) => data.add_model(Model{id, name}).await,  // get an error here
            Err(err) => Err(err.message.to_string()),
        }
    }
}
```

Is there any way to make data mutable? Or should I wrap everything in `Arc<RwLock>`?

Thank you in advance.

**Comment by janos-r** [comment]
I just wondered the exact same thing!
If I for example wanted to set headers to the response after the schema execution, based on the schema route, I guess this is the only place that could keep information from within the schema route.

**Comment by michalmoc** [comment]
As far as I know it is not currently possible. Context is expected to be used concurrently, and therefore it can return only non-mutable data. The proper way to access mutability is to add your own synchronization layer, e.g. storing `Arc<RwLock<your data>>` in context.

I personaly believe it should be possible to get &mut directly inside Mutations, but it is not implemented yet. I've raised it as feature request https://github.com/async-graphql/async-graphql/issues/1656, perhaps someday someone will manage to do it.
